src/boot/efi/boot.c

```diff
@@ -2357,6 +2357,9 @@ static EFI_STATUS image_start(
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Error preparing initrd: %m");
 
+        /* Ask shim to leave its protocol around, so that the stub can use it to validate PEs. */
+        shim_retain_protocol();
```

Comment:
i am not sure i grok what this means.

not knowing how this works with shim, but is it that the first validation we issue means shim unregisters its protocol for that? and that's what we want to disable here?

maybe say so in this comment "Normally, shim unregisters its protocol on the first PE validation attempt. Since we potentially want to validate multiple PE files however (i.e. UKI *and* multiple addon) we'll ask shim to stick around for good."

or something like that?

But did I even understand this properly? its just a guess how this works, you should know this better than i do.

btw, do you know *why* it deletes its protocol like this currently?

Reply:
Shim uninstalls the protocol once the kernel image is loaded (stage 3, I guess). This was done for historic reasons, to get out of the way as early as possible, but breaks this case as we are doing this from the stub, which is effectively the image. Adding an option was deemed better to avoid causing unforeseen issues.

Reply:
"once the image is loaded"? what does that mean? isn't this just done from their verification hook?

Reply:
"doing this from the stub"?

You are installing the variable from sd-boot, no?

Reply:
> To be honest, I also didn't get what this does. Apart from the _why_, I am pretty sure that current RHEL shim isn't version 15.8 (as @bluca wrote in a comment) but rather 15.6, and I successfully managed to test this. So are we sure we need this?

Yes, this does not work at all with upstream shim via sd-boot. Are you sure you didn't test loading directly the UKI or something? The issue is when going shim -> 2nd stage -> UKI stub

Reply:
hmm, looking at that link, what about the EFI_VARIABLE_BOOTSERVICE_ACCESS flag that that code checks for? don't we have to set that?

Reply:
> "doing this from the stub"?
> 
> You are installing the variable from sd-boot, no?

"doing this" -> loading and verifying the PE addons. We do it from the stub, and when going through shim -> sd-boot, due to the StartImage, the shim protocol is gone. With the change I've done in shim, and this variable, we can ensure the protocol is still available to the stub.

Reply:
anyway, tldr: please extend the comment explaining this in more detail, what we are doing here.

Reply:
already done

Reply:
> > To be honest, I also didn't get what this does. Apart from the _why_, I am pretty sure that current RHEL shim isn't version 15.8 (as @bluca wrote in a comment) but rather 15.6, and I successfully managed to test this. So are we sure we need this?
> 
> Yes, this does not work at all with upstream shim via sd-boot. Are you sure you didn't test loading directly the UKI or something? The issue is when going shim -> 2nd stage -> UKI stub

Yes, I actually loaded directly the UKI. All good, thanks